### PR TITLE
Promote serving cert annotations to GA

### PR DESCRIPTION
**Pull request by smarterclayton**
No transition in the controller. Instead, implement

    oc adm migrate alpha serving-certs

which copies the annotations to their canonical name. Can be run pre or post migration. Avoids complex logic in the controller.

@deads2k @enj @liggitt

### CODE REVIEW

**Comment by deads2k** [comment]
I'd really like to try to sort out a flex volume plugin or something to mount the CA bundle before committing to GA

**Comment by smarterclayton** [comment]
Why is that a prereq?  You're worried that the annotations on the backend
will change?

On Sep 8, 2017, at 7:56 AM, David Eads <notifications@github.com> wrote:

I'd really like to try to sort out a flex volume plugin or something to
mount the CA bundle before committing to GA

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/16225#issuecomment-328083962>,
or mute
the thread

**Comment by smarterclayton** [comment]
Re: GA we are dependent on these annotations for improving components.
What sorts of discoveries would we make with flex volume that would cause
us to change the implementation here?

On Sep 8, 2017, at 7:56 AM, David Eads <notifications@github.com> wrote:

I'd really like to try to sort out a flex volume plugin or something to
mount the CA bundle before committing to GA

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/16225#issuecomment-328083962>,
or mute
the thread

**Comment by deads2k** [comment]
> Re: GA we are dependent on these annotations for improving components.
> What sorts of discoveries would we make with flex volume that would cause
> us to change the implementation here?

Just that there are two halves to promoting this.  The "make my secret" half and the "how do I trust you" half.  It seems like we should promote as a pair with a release or so of cut-over time.

**Comment by enj** [comment]
@openshift/sig-security

File: pkg/oc/admin/migrate/servingcert/servingcert.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+		}
+		var err error
+		if changed && !o.ResourceOptions.DryRun {
+			_, err = resource.NewHelper(info.Client, info.Mapping).Replace(info.Namespace, info.Name, false, info.Object)
+		}
+		return migrate.ReporterBool(changed), err
+	default:
+		return nil, nil // indicate that we ignored the object
+	}
+	return migrate.NotChanged, nil
```
Comment (enj) [inline pkg/oc/admin/migrate/servingcert/servingcert.go]: Can you even reach this line?

File: pkg/oc/admin/migrate/migrate.go [hunk 2]
```diff
@@ -28,3 +31,15 @@ func NewCommandMigrate(name, fullName string, f *clientcmd.Factory, out, errOut
 	cmd.AddCommand(cmds...)
 	return cmd
 }
+
+func NewCommandMigrateAlpha(name, fullName string, f *clientcmd.Factory, out, errOut io.Writer, cmds ...*cobra.Command) *cobra.Command {
```
Comment (enj) [inline pkg/oc/admin/migrate/migrate.go]: I feel like we are starting to abuse the migrations we do via ansible.

**Comment by liggitt** [comment]
> Re: GA we are dependent on these annotations for improving components. What sorts of discoveries would we make with flex volume that would cause us to change the implementation here?

I think piggy-backing on the service account token secret was a mistake we should rectify before exiting alpha.

File: pkg/oc/admin/migrate/migrate.go (hunk 2 above)
Comment (smarterclayton) [reply pkg/oc/admin/migrate/migrate.go]: Why?

Comment (smarterclayton) [reply pkg/oc/admin/migrate/migrate.go]: This wouldn't be for ansible.  This is a convenience for someone who wants to move alpha annotations forward without having to write the script themselves.

**Pull request closed without merging**